src/__tests__/example-email-advanced.ts

```diff
@@ -0,0 +1,90 @@
+import {
+  anyOf,
+  buildRegExp,
+  charClass,
+  charRange,
+  digit,
+  endOfString,
+  namedCapture,
+  oneOrMore,
+  repeat,
+  startOfString,
+} from '..';
+
+//
+// Example: email validation building blocks
+//
+const upperCase = charRange('A', 'Z');
+const lowerCase = charRange('a', 'z');
+const specialChars = anyOf("!#$%&'*+/=?^_`{|}~-");
+const usernameChars = charClass(upperCase, lowerCase, digit, specialChars);
+const hostnameChars = charClass(upperCase, lowerCase, digit, specialChars);
+const domainChars = charRange('a', 'z');
+const emailSeparator = anyOf('.');
+const domainSeparator = anyOf('@');
+
+//
+// Example: email validation majour components using named capture.
+//
+const username = namedCapture(oneOrMore(usernameChars), 'username');
+
+const usernameRegex = buildRegExp([startOfString, username, endOfString]);
+
+test('Matching the Username component.', () => {
```

Comment:
I'll merge this with existing email example, as there quite similar. 

Note: I am planning to add some frequently used patterns (URL, email, maybe hashtags, etc). So that each user does not have to define them by hand. I will soon spec-out this feature. I invite you to join in if you have capacity for that.